### PR TITLE
AOT changes

### PR DESCRIPTION
**Pull request by veresov**
This is a all of the AOT changes. Includes constant replacement and resolution machinery, tiered compilation support (with probabilistic profiling).

### CODE REVIEW

File: graal/com.oracle.graal.hotspot.amd64/src/com/oracle/graal/hotspot/amd64/AMD64HotSpotBackendFactory.java [hunk 1]
```diff
@@ -187,8 +188,7 @@ protected AMD64HotSpotForeignCallsProvider createForeignCalls(HotSpotJVMCIRuntim
      */
     protected HotSpotSuitesProvider createSuites(GraalHotSpotVMConfig config, HotSpotGraalRuntimeProvider runtime, CompilerConfiguration compilerConfiguration, Plugins plugins,
                     HotSpotRegistersProvider registers, Replacements replacements) {
-        return new HotSpotSuitesProvider(new AMD64HotSpotSuitesProvider(compilerConfiguration, plugins), config, runtime, new AMD64HotSpotAddressLowering(config.getOopEncoding().base,
-                        registers.getHeapBaseRegister()));
+        return new HotSpotSuitesProvider(new AMD64SuitesProvider(compilerConfiguration, plugins), config, runtime, new AMD64HotSpotAddressLowering(config, registers.getHeapBaseRegister()));
```
Comment (dougxc) [inline graal/com.oracle.graal.hotspot.amd64/src/com/oracle/graal/hotspot/amd64/AMD64HotSpotBackendFactory.java]: Why this change?

**Comment by zapster** [comment]
A few minor issues that should be addressed at some point. They are not really crucial for acceptance.

* Some of the new `Nodes` and `LIRInstructions` are not really self-explanatory and could use javadoc (e.g. `AMD64HotSpotConstantRetrievalOp`)
* Multi-line inline comments should use block comments (`/* */`) instead of `//` to avoid formatting issues during refactoring.
* Would it be possible to add tests for the new features without relying on an AOT enabled JDK?

**Comment by dougxc** [comment]
Also to be fixed later:
* Need documentation for `ProfilingPlugin` and `ClassInitializationPlugin`

**Comment by veresov** [comment]
Thanks for the comments guys, I'll make sure add more documentation in the next pass.
The absence of the unit tests is a very good point. This is definitely something that we need to address after the JDK9 rush ends.

File: graal/com.oracle.graal.java/src/com/oracle/graal/java/BytecodeParser.java [hunk 2]
```diff
@@ -3251,8 +3305,12 @@ private Object lookupConstant(int cpi, int opcode) {
     }
 
     private void maybeEagerlyResolve(int cpi, int bytecode) {
-        if (graphBuilderConfig.eagerResolving() || intrinsicContext != null) {
+        if (intrinsicContext != null) {
             constantPool.loadReferencedType(cpi, bytecode);
+        } else if (graphBuilderConfig.eagerResolving()) {
+            synchronized (BytecodeParser.class) {
```
Comment (dougxc) [inline graal/com.oracle.graal.java/src/com/oracle/graal/java/BytecodeParser.java]: What is this for? @tkrodriguez mentioned vaguely remember fixing a synchronization issue in this area.

Comment (veresov) [inline graal/com.oracle.graal.java/src/com/oracle/graal/java/BytecodeParser.java]: I remember adding this, but I don't remember what was the issue exactly. Let me remove and see what happens.

Comment (veresov) [inline graal/com.oracle.graal.java/src/com/oracle/graal/java/BytecodeParser.java]: It don't seem to cause any trouble. Perhaps this problem had been fixed elsewhere. I think we can remove this change.

Comment (veresov) [inline graal/com.oracle.graal.java/src/com/oracle/graal/java/BytecodeParser.java]: Seems to be required because running some static initializers in the JDK concurrently causes deadlocks.

Comment (veresov) [inline graal/com.oracle.graal.java/src/com/oracle/graal/java/BytecodeParser.java]: Here's how the deadlock looks like:

"pool-1-thread-5" #16 prio=5 os_prio=0 tid=0x00007f5570787800 nid=0x22f9 in Object.wait() [0x00007f5400ada000]
   java.lang.Thread.State: RUNNABLE
	at sun.security.provider.SeedGenerator.<clinit>(java.base@9-internal/SeedGenerator.java:105)
	at jdk.internal.misc.Unsafe.ensureClassInitialized0(java.base@9-internal/Native Method)
	at jdk.internal.misc.Unsafe.ensureClassInitialized(java.base@9-internal/Unsafe.java:1023)
	at jdk.vm.ci.hotspot.HotSpotConstantPool.loadReferencedType(jdk.vm.ci@9-internal/HotSpotConstantPool.java:696)
	at com.oracle.graal.java.BytecodeParser.maybeEagerlyResolve(jdk.vm.compiler@9-internal/BytecodeParser.java:3312)
	at com.oracle.graal.java.BytecodeParser.lookupMethod(jdk.vm.compiler@9-internal/BytecodeParser.java:3272)
	at com.oracle.graal.java.BytecodeParser.processBytecode(jdk.vm.compiler@9-internal/BytecodeParser.java:4019)
	at com.oracle.graal.java.BytecodeParser.iterateBytecodesForBlock(jdk.vm.compiler@9-internal/BytecodeParser.java:2624)
	at com.oracle.graal.java.BytecodeParser.processBlock(jdk.vm.compiler@9-internal/BytecodeParser.java:2441)
	at com.oracle.graal.java.BytecodeParser.build(jdk.vm.compiler@9-internal/BytecodeParser.java:770)
	at com.oracle.graal.java.BytecodeParser.buildRootMethod(jdk.vm.compiler@9-internal/BytecodeParser.java:668)
	at com.oracle.graal.java.GraphBuilderPhase$Instance.run(jdk.vm.compiler@9-internal/GraphBuilderPhase.java:93)
	at com.oracle.graal.java.GraphBuilderPhase.run(jdk.vm.compiler@9-internal/GraphBuilderPhase.java:57)
	at com.oracle.graal.java.GraphBuilderPhase.run(jdk.vm.compiler@9-internal/GraphBuilderPhase.java:41)
	at com.oracle.graal.phases.BasePhase.apply(jdk.vm.compiler@9-internal/BasePhase.java:149)
	at com.oracle.graal.phases.BasePhase.apply(jdk.vm.compiler@9-internal/BasePhase.java:131)
	at com.oracle.graal.phases.PhaseSuite.run(jdk.vm.compiler@9-internal/PhaseSuite.java:154)
	at com.oracle.graal.phases.BasePhase.apply(jdk.vm.compiler@9-internal/BasePhase.java:149)
	at com.oracle.graal.phases.BasePhase.apply(jdk.vm.compiler@9-internal/BasePhase.java:131)
	at jdk.tools.jaotc.AOTBackend.buildStructuredGraph(jdk.aot@9-internal/AOTBackend.java:115)
	at jdk.tools.jaotc.AOTBackend.compileMethod(jdk.aot@9-internal/AOTBackend.java:97)
	at jdk.tools.jaotc.AOTCompilationTask.run(jdk.aot@9-internal/AOTCompilationTask.java:117)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@9-internal/ThreadPoolExecutor.java:1161)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@9-internal/ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(java.base@9-internal/Thread.java:844)

"pool-1-thread-4" #15 prio=5 os_prio=0 tid=0x00007f5570785800 nid=0x22f8 in Object.wait() [0x00007f5400bdb000]
   java.lang.Thread.State: RUNNABLE
	at sun.security.provider.SecureRandom$SeederHolder.<clinit>(java.base@9-internal/SecureRandom.java:202)
	at jdk.internal.misc.Unsafe.ensureClassInitialized0(java.base@9-internal/Native Method)
	at jdk.internal.misc.Unsafe.ensureClassInitialized(java.base@9-internal/Unsafe.java:1023)
	at jdk.vm.ci.hotspot.HotSpotConstantPool.loadReferencedType(jdk.vm.ci@9-internal/HotSpotConstantPool.java:696)
	at com.oracle.graal.java.BytecodeParser.maybeEagerlyResolve(jdk.vm.compiler@9-internal/BytecodeParser.java:3312)
	at com.oracle.graal.java.BytecodeParser.lookupField(jdk.vm.compiler@9-internal/BytecodeParser.java:3286)
	at com.oracle.graal.java.BytecodeParser.processBytecode(jdk.vm.compiler@9-internal/BytecodeParser.java:4014)
	at com.oracle.graal.java.BytecodeParser.iterateBytecodesForBlock(jdk.vm.compiler@9-internal/BytecodeParser.java:2624)
	at com.oracle.graal.java.BytecodeParser.processBlock(jdk.vm.compiler@9-internal/BytecodeParser.java:2441)
	at com.oracle.graal.java.BytecodeParser.build(jdk.vm.compiler@9-internal/BytecodeParser.java:770)
	at com.oracle.graal.java.BytecodeParser.buildRootMethod(jdk.vm.compiler@9-internal/BytecodeParser.java:668)
	at com.oracle.graal.java.GraphBuilderPhase$Instance.run(jdk.vm.compiler@9-internal/GraphBuilderPhase.java:93)
	at com.oracle.graal.java.GraphBuilderPhase.run(jdk.vm.compiler@9-internal/GraphBuilderPhase.java:57)
	at com.oracle.graal.java.GraphBuilderPhase.run(jdk.vm.compiler@9-internal/GraphBuilderPhase.java:41)
	at com.oracle.graal.phases.BasePhase.apply(jdk.vm.compiler@9-internal/BasePhase.java:149)
	at com.oracle.graal.phases.BasePhase.apply(jdk.vm.compiler@9-internal/BasePhase.java:131)
	at com.oracle.graal.phases.PhaseSuite.run(jdk.vm.compiler@9-internal/PhaseSuite.java:154)
	at com.oracle.graal.phases.BasePhase.apply(jdk.vm.compiler@9-internal/BasePhase.java:149)
	at com.oracle.graal.phases.BasePhase.apply(jdk.vm.compiler@9-internal/BasePhase.java:131)
	at jdk.tools.jaotc.AOTBackend.buildStructuredGraph(jdk.aot@9-internal/AOTBackend.java:115)
	at jdk.tools.jaotc.AOTBackend.compileMethod(jdk.aot@9-internal/AOTBackend.java:97)
	at jdk.tools.jaotc.AOTCompilationTask.run(jdk.aot@9-internal/AOTCompilationTask.java:117)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@9-internal/ThreadPoolExecutor.java:1161)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@9-internal/ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(java.base@9-internal/Thread.java:844)

"pool-1-thread-3" #14 prio=5 os_prio=0 tid=0x00007f5570783800 nid=0x22f7 in Object.wait() [0x00007f5400cdc000]
   java.lang.Thread.State: RUNNABLE
	at jdk.internal.misc.Unsafe.ensureClassInitialized0(java.base@9-internal/Native Method)
	at jdk.internal.misc.Unsafe.ensureClassInitialized(java.base@9-internal/Unsafe.java:1023)
	at jdk.vm.ci.hotspot.HotSpotConstantPool.loadReferencedType(jdk.vm.ci@9-internal/HotSpotConstantPool.java:696)
	at com.oracle.graal.java.BytecodeParser.maybeEagerlyResolve(jdk.vm.compiler@9-internal/BytecodeParser.java:3312)
	at com.oracle.graal.java.BytecodeParser.lookupField(jdk.vm.compiler@9-internal/BytecodeParser.java:3286)
	at com.oracle.graal.java.BytecodeParser.processBytecode(jdk.vm.compiler@9-internal/BytecodeParser.java:4016)
	at com.oracle.graal.java.BytecodeParser.iterateBytecodesForBlock(jdk.vm.compiler@9-internal/BytecodeParser.java:2624)
	at com.oracle.graal.java.BytecodeParser.processBlock(jdk.vm.compiler@9-internal/BytecodeParser.java:2441)
	at com.oracle.graal.java.BytecodeParser.build(jdk.vm.compiler@9-internal/BytecodeParser.java:770)
	at com.oracle.graal.java.BytecodeParser.buildRootMethod(jdk.vm.compiler@9-internal/BytecodeParser.java:668)
	at com.oracle.graal.java.GraphBuilderPhase$Instance.run(jdk.vm.compiler@9-internal/GraphBuilderPhase.java:93)
	at com.oracle.graal.java.GraphBuilderPhase.run(jdk.vm.compiler@9-internal/GraphBuilderPhase.java:57)
	at com.oracle.graal.java.GraphBuilderPhase.run(jdk.vm.compiler@9-internal/GraphBuilderPhase.java:41)
	at com.oracle.graal.phases.BasePhase.apply(jdk.vm.compiler@9-internal/BasePhase.java:149)
	at com.oracle.graal.phases.BasePhase.apply(jdk.vm.compiler@9-internal/BasePhase.java:131)
	at com.oracle.graal.phases.PhaseSuite.run(jdk.vm.compiler@9-internal/PhaseSuite.java:154)
	at com.oracle.graal.phases.BasePhase.apply(jdk.vm.compiler@9-internal/BasePhase.java:149)
	at com.oracle.graal.phases.BasePhase.apply(jdk.vm.compiler@9-internal/BasePhase.java:131)
	at jdk.tools.jaotc.AOTBackend.buildStructuredGraph(jdk.aot@9-internal/AOTBackend.java:115)
	at jdk.tools.jaotc.AOTBackend.compileMethod(jdk.aot@9-internal/AOTBackend.java:97)
	at jdk.tools.jaotc.AOTCompilationTask.run(jdk.aot@9-internal/AOTCompilationTask.java:117)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@9-internal/ThreadPoolExecutor.java:1161)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@9-internal/ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(java.base@9-internal/Thread.java:844)

Comment (tkrodriguez) [inline graal/com.oracle.graal.java/src/com/oracle/graal/java/BytecodeParser.java]: Eager resolution is generally exposed to this sort of problem, though of course running multiple initializations makes this worse.  A comment explaining that this helps avoid deadlocks seems like a good idea since I don't see a better "solution" to this.

File: graal/com.oracle.graal.phases.common/src/com/oracle/graal/phases/common/inlining/walker/InliningData.java [hunk 3]
```diff
@@ -218,15 +218,19 @@ private InlineInfo getInlineInfo(Invoke invoke) {
...
-                ResolvedJavaMethod resolvedMethod = leafConcreteSubtype.getResult().resolveConcreteMethod(targetMethod, contextType);
-                if (resolvedMethod != null) {
+        AssumptionResult<ResolvedJavaType> leafConcreteSubtype = holder.findLeafConcreteSubtype();
+        if (leafConcreteSubtype != null) {
+            ResolvedJavaMethod resolvedMethod = leafConcreteSubtype.getResult().resolveConcreteMethod(targetMethod, contextType);
+            if (resolvedMethod != null) {
+                if (callTarget.graph().getAssumptions() != null) {
                     return getAssumptionInlineInfo(invoke, resolvedMethod, leafConcreteSubtype);
+                } else {
+                    return getTypeCheckedAssumptionInfo(invoke, resolvedMethod, leafConcreteSubtype.getResult());
```
Comment (tkrodriguez) [inline graal/com.oracle.graal.phases.common/src/com/oracle/graal/phases/common/inlining/walker/InliningData.java]: This seems like a dubious thing to have on all the time, though maybe it works out ok?  I'm not sure what other contexts this would be exercised besides AOT.  Maybe it should be GenericPIC guarded?

File: graal/com.oracle.graal.java/src/com/oracle/graal/java/BytecodeParser.java (hunk 2 above)
Comment (dougxc) [inline graal/com.oracle.graal.java/src/com/oracle/graal/java/BytecodeParser.java]: I added a comment.

File: graal/com.oracle.graal.hotspot.amd64/src/com/oracle/graal/hotspot/amd64/AMD64HotSpotLoadConfigValueOp.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @Override
+    public void emitCode(CompilationResultBuilder crb, AMD64MacroAssembler masm) {
+        if (GeneratePIC.getValue()) {
+            masm.movq(asRegister(result), masm.getPlaceholder(-1));
+        } else {
+            throw GraalError.unimplemented();
+        }
+        crb.recordMark(markId);
```
Comment (rschatz) [inline graal/com.oracle.graal.hotspot.amd64/src/com/oracle/graal/hotspot/amd64/AMD64HotSpotLoadConfigValueOp.java]: Why is the `mark` recorded at the *next* instruction, and not directly on the `movq`?

File: graal/com.oracle.graal.hotspot/src/com/oracle/graal/hotspot/FingerprintUtil.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import jdk.vm.ci.hotspot.HotSpotResolvedObjectType;
+
+/**
+ * Reflective wrapper around the fingerprint generation functionality in JDK9.
+ */
+public class FingerprintUtil {
+    private static Method getFingerprint;
+    static {
+        try {
+            getFingerprint = HotSpotResolvedObjectType.class.getMethod("getFingerprint");
```
Comment (rschatz) [inline graal/com.oracle.graal.hotspot/src/com/oracle/graal/hotspot/FingerprintUtil.java]: Can't we just add a dummy method to `graal-jvmci-8` for compatibility?

Comment (dougxc) [inline graal/com.oracle.graal.hotspot/src/com/oracle/graal/hotspot/FingerprintUtil.java]: We could but I don't want to wait for a JVMCI-8 release cycle before merging this PR. Also, I don't see this method in [HotSpotResolvedObjectType](http://hg.openjdk.java.net/jdk9/hs/hotspot/file/tip/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotResolvedObjectType.java) in JDK9?

Comment (dougxc) [inline graal/com.oracle.graal.hotspot/src/com/oracle/graal/hotspot/FingerprintUtil.java]: I believe it must be the latter - can you please confirm @veresov .

**Comment by dougxc** [comment]
These changes have now been merged: https://github.com/graalvm/graal-core/commit/2a6ad157219d37f0118f9a17624d35abe6508a67

File: graal/com.oracle.graal.hotspot.amd64/src/com/oracle/graal/hotspot/amd64/AMD64HotSpotLoadConfigValueOp.java (hunk 4 above)
Comment (dougxc) [inline graal/com.oracle.graal.hotspot.amd64/src/com/oracle/graal/hotspot/amd64/AMD64HotSpotLoadConfigValueOp.java]: @veresov  , if this is a bug, please submit a new PR to fix it.

File: graal/com.oracle.graal.phases.common/src/com/oracle/graal/phases/common/inlining/walker/InliningData.java (hunk 3 above)
Comment (veresov) [inline graal/com.oracle.graal.phases.common/src/com/oracle/graal/phases/common/inlining/walker/InliningData.java]: We could, but it's already guarded by assumptions==null. Which is only used by AOT as far as I know.

File: graal/com.oracle.graal.hotspot/src/com/oracle/graal/hotspot/FingerprintUtil.java (hunk 5 above)
Comment (veresov) [inline graal/com.oracle.graal.hotspot/src/com/oracle/graal/hotspot/FingerprintUtil.java]: It's will be in the VM changes for AOT which are going to be pushed later. Propagating fingerprint support to JDK8 would be pretty hard as that would require extensive VM changes and that won't be approved. We can, later, add getFingerprint() to JVMCI8 that would just return 0 and get rid of the wrapper. But that will have to be done after the rest of AOT is pushed.

Comment (rschatz) [inline graal/com.oracle.graal.hotspot/src/com/oracle/graal/hotspot/FingerprintUtil.java]: Yes, I meant including a dummy method that just returns 0.

File: graal/com.oracle.graal.hotspot.amd64/src/com/oracle/graal/hotspot/amd64/AMD64HotSpotLoadConfigValueOp.java (hunk 4 above)
Comment (veresov) [inline graal/com.oracle.graal.hotspot.amd64/src/com/oracle/graal/hotspot/amd64/AMD64HotSpotLoadConfigValueOp.java]: I think it's mostly historic. I think the decision to put the mark on the next instruction is because on x86 the address is at the end of the instruction. So to patch it you don't have to decode the instruction, you would just need to do mark_address - 4. I agree that a bit unintuitive and awkward, and a bad idea for other architectures. We'd need to clean that up in the future.

Comment (rschatz) [inline graal/com.oracle.graal.hotspot.amd64/src/com/oracle/graal/hotspot/amd64/AMD64HotSpotLoadConfigValueOp.java]: Ok, makes sense. A comment wouldn't hurt, though, to avoid confusing everyone else stumbling over that code ;)

File: graal/com.oracle.graal.hotspot/src/com/oracle/graal/hotspot/FingerprintUtil.java (hunk 5 above)
Comment (dougxc) [inline graal/com.oracle.graal.hotspot/src/com/oracle/graal/hotspot/FingerprintUtil.java]: I've pushed the [change](http://hg.openjdk.java.net/graal/graal-jvmci-8/rev/f272ef93c3a3) to jvmci-8 and we can use it once jvmci-0.24 is released.

Comment (veresov) [inline graal/com.oracle.graal.hotspot/src/com/oracle/graal/hotspot/FingerprintUtil.java]: Perfect! Thanks, @dougxc !
